### PR TITLE
Exercise 5

### PR DESCRIPTION
**Pull request by Cliver67**
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

### CODE REVIEW

**Comment by Cliver67** [comment]
Am looking for feedback , build commands seem to work...  fully expect comments and reworking

File: .env.prd
Comment (hugh-emerson) [inline .env.prd]: You've accidentally leaked credentials.

Delete this file (and if you care about your Trello account at all, rotate the key)

If you want to keep a copy locally for whatever reason (though the .env file should suffice), then you can add .env.prd to your .gitignore file and run `git rm --cached .env.prd` in your terminal.

You didn't add it to your .dockerignore either so it's currently getting included in the `COPY` command in your Dockerfile. We do not want any credentials baked into the image.

File: .env.prd
Comment (hugh-emerson) [inline .env.prd]: Note that the .env files are not Python. The correct way to write this is `BOARDID=608a91fd8a30ef4966d68da1`

python-dotenv strips the spaces and quotes but if you want to run `docker run --env-file .env my-image` then you need to follow that format strictly.

File: README.md
Comment (hugh-emerson) [inline README.md]: You should document the fact that you can run the app in Docker / what commands to run.

File: Dockerfile
Comment (hugh-emerson) [inline Dockerfile]: I think you mean `COPY . ./`

If the path starts with a / that will copy things to the root of the Linux system. You want to copy into the current working directory.

But you would have seen the app still working because of the bind mount

File: Docker-compose.yml
Comment (hugh-emerson) [inline Docker-compose.yml]: This should be lowercase. Same for `Production` below

File: Docker-compose.yml
Comment (hugh-emerson) [inline Docker-compose.yml]:
```suggestion
    volumes: 
```

Same for the one below.

File: Docker-compose.yml
Comment (hugh-emerson) [inline Docker-compose.yml:17]: You need to publish a port if you want to actually visit it in your browser. E.g:

```
    ports:
      - "5000:5000"
```

(Equivalent to the docker command line's `-p` option)

File: Docker-compose.yml
Comment (hugh-emerson) [inline Docker-compose.yml]: You need to specify which stage you want to build. If you don't, it'll build the lowest one in the Dockerfile.

You need to expand this build section:
```
    build: 
      context: .
      target: development
```

File: Docker-compose.yml
Comment (hugh-emerson) [inline Docker-compose.yml]:
```suggestion
          source: ./todo_app
          target: /todo_app/todo_app
```

Don't forget the `.` to mean relative to the current directory and you want to keep the file structure - python files in a todo_app subfolder. This copies the contents of the source folder into the contents of the target folder. So you need to keep the second `/todo_app` 

Not sure where `/app` came from - you set WORKDIR to `/todo_app` in your Dockerfile.

File: Dockerfile
Comment (hugh-emerson) [inline Dockerfile]: You need --host 0.0.0.0 in your flask run command for it to listen to external requests. Same idea as including --bind 0.0.0.0 in your gunicorn command.

**Review by hugh-emerson — CHANGES_REQUESTED** [review]
You're mostly there but some issues need to be fixed for it to run plus you've leaked your credentials . See my comments. Getting the compose file working is a stretch goal but I've given some guidance with that. If you get the "docker build" and "docker run" commands working, then document them in the README.md

It definitely doesn't work as-is, so if you thought you saw it working then perhaps you were actually visiting a copy of the app that was running outside of Docker? Or perhaps you have some changes that aren't pushed to GitHub.

Also, aside from the Docker work: 
- You've accidentally added a `DevOps-Course-Workshop-Module-07-Learners` submodule. That should be removed.
- Please merge approved pull requests into the main branch - either manually in your terminal (and push it) or click the merge button in GitHub at the bottom of the PR

File: Dockerfile
Comment (Cliver67) [reply Dockerfile]: change made

File: Dockerfile
Comment (Cliver67) [reply Dockerfile]: amendment made

File: Docker-compose.yml
Comment (Cliver67) [reply Docker-compose.yml]: change made

File: README.md
Comment (Cliver67) [reply README.md]: build and run commands documented in the readme

**Comment by Cliver67** [comment]
re merging , ex 4 and 5 branches were generated off ex 3 instead of off Main..

plan is to complete 3 and merge all together (if i ever get 3 completed~)

ex 6 was taken from main , so once reviewed and completed can be merged back

File: Dockerfile
Comment (JackMeadDev) [inline Dockerfile:22]: Typo here that will cause you problems!

```suggestion
CMD ["poetry", "run", "flask", "run", "--host",  "0.0.0.0"]    
```

File: todo_app/data/trello_items.py
Comment (JackMeadDev) [inline todo_app/data/trello_items.py]: You're not actually calling this function here, so it's trying to stringify the function itself rather than the result of it - leading to JSON decoding errors when Trello denies know which board you're referring to
```suggestion
    url = 'https://api.trello.com/1/boards/' + str(get_boardid()) +'/cards?fields=name,idList,dateLastActivity'
```

File: todo_app/data/trello_items.py
Comment (JackMeadDev) [inline todo_app/data/trello_items.py:52]: Similarly here, and various other places below
```suggestion
        'key': str(get_trello_key()),
        'token': str(get_trello_token())
```

File: Docker-compose.yml
Comment (JackMeadDev) [inline Docker-compose.yml:25]: You fixed up the other one, but this still has a `.` in front of that causes it to error

File: Docker-compose.yml
Comment (JackMeadDev) [inline Docker-compose.yml:31]: Worth noting that with both apps pointing to port 5000 on the host, only one can be run at a time. That's fine if you want people to run them one at a time (e.g. with `docker-compose up production`) but we could switch this to an alternative port so that both can be run at the same time - moving the prod setup to the below would mean we leave dev available on your computer's 5000 while prod is available on 5001
```suggestion
        - "5001:5000" 
```

File: README.md
Comment (JackMeadDev) [inline README.md:101]: We should add in the config for hot reloading so that development can easily take place without constantly restarting the container;
```suggestion
docker run --rm -it -p 5000:5000/tcp --env-file .env --mount type=bind,source="$(pwd)"/todo_app,target=/todo_app/todo_app todo-app:dev 
```

**Review by JackMeadDev — CHANGES_REQUESTED** [review]
Thanks for making those changes, and nice work getting your docker-compose together too, but there were still a few issues that prevented me from running your containers successfully following the instructions - I've left some thoughts below for fixing those up. Feel free to reach out if any of it is unclear!

File: README.md
Comment (Cliver67) [reply README.md:101]: thank you  - leaving both in the read me to allow options

**Comment by Cliver67** [comment]
> You're mostly there but some issues need to be fixed for it to run plus you've leaked your credentials . See my comments. Getting the compose file working is a stretch goal but I've given some guidance with that. If you get the "docker build" and "docker run" commands working, then document them in the README.md
> 
> It definitely doesn't work as-is, so if you thought you saw it working then perhaps you were actually visiting a copy of the app that was running outside of Docker? Or perhaps you have some changes that aren't pushed to GitHub.
> 
> Also, aside from the Docker work:
> 
> * You've accidentally added a `DevOps-Course-Workshop-Module-07-Learners` submodule. That should be removed.
> * Please merge approved pull requests into the main branch - either manually in your terminal (and push it) or click the merge button in GitHub at the bottom of the PR

accidental inclusion or module 7 learner sub module has been addressed and removed

**Comment by JackMeadDev** [comment]
PR has been resubmitted under https://github.com/Cliver67/DevOps-Course-Starter/pull/9 - it should be reviewed there

**Pull request closed without merging**
